### PR TITLE
feat: add Go samples

### PR DESCRIPTION
**Pull request by enocom**

### CODE REVIEW

**Comment by enocom** [comment]
cc @codyoss. Here's a first pass adding support for Go. I've tried to follow the Go contributing guidelines here, but there will still be more work to bring the two guides together.

File: content/en/_index.md [hunk 1]
```diff
@@ -80,14 +80,24 @@ def example_snippet():
 
 # [END product_example]
 {{< /tab >}}
+{{<tab header="Go">}}
+# [START product_example]
+import "example.com/resource"
+
+func exampleSnippet() {
+    // Snippet Content ...
+}
```
Comment (kurtisvg) [inline content/en/_index.md:89]: nit: consider using this to show it should go after the package, but before the imports
```suggestion
package main;

// Region tags should start after the package, but before imports.
// [START product_example]
import "example.com/resource"
// [END product_example]
```

File: content/en/_index.md [hunk 2]
```diff
@@ -250,6 +339,12 @@ try (CloudClient dlp = CloudClient.create()) {
   // make a request with the client
 }
 {{< /tab >}}
+{{< tab header="Go" >}}
+// Don't initialize one client for the entire set of samples and pass it as an
+// argument. Each sample should initialize its own client/service.
```
Comment (kurtisvg) [inline content/en/_index.md:344]: This comment should be targeted to the user and should try give guidance on how to use clients (and if there are any limits - e.g. thread safety), and if there is any clean up needed to remove background resources.

File: content/en/_index.md [hunk 3]
```diff
@@ -197,6 +236,50 @@ public static void listInfoTypes() throws IOException {
...
+{{< tab header="Go" >}}
+import (
+	"context"
+	"fmt"
+
+	dlp "cloud.google.com/go/dlp/apiv2"
+	dlppb "google.golang.org/genproto/googleapis/privacy/dlp/v2"
+)
+
+// listInfoTypes lists the types of sensitive information the DLP API supports.
```
Comment (kurtisvg) [inline content/en/_index.md:248]: Should this have the `main()` function here?

File: content/en/_index.md [hunk 4]
```diff
@@ -148,6 +179,14 @@ public static void exampleSnippet(String projectId, String filePath) {
     // Snippet content ...
 }
 {{< /tab >}}
+{{< tab header="Go" >}}
+// exampleSnippet demonstrates best practices.
+// The function prints to an io.Writer for testing purposes. Do not print to
+// stdout or stderr. Use the io.Writer instead.
+func exampleSnippet(w io.Writer, projectId, filePath string) error {
+    // Snippet content ...
+}
```
Comment (kurtisvg) [inline content/en/_index.md:188]: I think using an error return type here goes against the guideline:

> Methods should avoid a return type whenever possible. Instead, show the user how
to interact with a returned object programmatically by printing some example
attributes to the console.

**Review by kurtisvg — CHANGES_REQUESTED** [review]

File: content/en/_index.md (hunk 4 above)
Comment (enocom) [reply content/en/_index.md:188]: This might be worth updating, given that most other languages here (Java, Python, Node) will use exceptions. Granted, we could write an error to the io.Writer, but for Go that's pretty unusual.

Comment (kurtisvg) [reply content/en/_index.md:188]: This is actually intentional, the goal is to force the sample to write what went wrong rather than rely on the test to catch errors. [Here's a similar example in Java](https://github.com/GoogleCloudPlatform/java-docs-samples/blob/9f6a4731b112e3554bc245abb0836b81e5acaf7c/iam/api-client/src/main/java/iam/snippets/DisableServiceAccount.java#L56).

Comment (enocom) [reply content/en/_index.md:188]: So the idea is that both success and failure are reported to the console. I like the symmetry.

File: content/en/_index.md (hunk 3 above)
Comment (enocom) [reply content/en/_index.md:248]: I've matched the Java usage above. I think using a non-main function makes it more copy-paste friendly.

File: content/en/_index.md (hunk 1 above)
Comment (enocom) [reply content/en/_index.md:92]: Added a package above.

File: content/en/_index.md (hunk 2 above)
Comment (enocom) [reply content/en/_index.md]: Fixed.

**Review by kurtisvg — APPROVED** [review]

File: content/en/_index.md (hunk 4 above)
Comment (codyoss) [reply content/en/_index.md]: I am not sure I agree with this for Go. We would be forced to log.Fatal in samples, which is a panic and anti-pattern. I understand the proposition for exception based languages, but I don't think this rule translates well for Go.

File: content/en/_index.md [hunk 5]
```diff
@@ -115,6 +134,14 @@ public static void main(String[] args) {
     exampleSnippet(projectId, filePath);
 }
 {{< /tab >}}
+{{< tab header="Go" >}}
+func main() {
+    // TODO(developer): Replace these variables before running the sample.
+    projectId := "my-project-id"
+    filePath := "path/to/image.png"
+    exampleSnippet(projectId, filePath)
```
Comment (codyoss) [inline content/en/_index.md:142]: If we are requiring a main running I would much perfer to see errors handled here for Go.

File: content/en/_index.md [hunk 6]
```diff
@@ -250,6 +341,14 @@ try (CloudClient dlp = CloudClient.create()) {
   // make a request with the client
 }
 {{< /tab >}}
+{{< tab header="Go" >}}
+// Initialize client that will be used to send requests. This client only needs
+// to be created once, and can be reused for multiple requests. After completing
+// all of your requests, call the "Close" method on the client to safely clean
+// up any remaining background resources.
+ctx := context.Background()
+c, err := foo.NewClient(ctx)
```
Comment (codyoss) [inline content/en/_index.md:350]: show checking the error and deferring the client Close

File: content/en/_index.md [hunk 7]
```diff
@@ -148,6 +181,14 @@ public static void exampleSnippet(String projectId, String filePath) {
     // Snippet content ...
 }
 {{< /tab >}}
+{{< tab header="Go" >}}
+// exampleSnippet demonstrates best practices.
+// The function prints to an io.Writer for testing purposes. Do not print to
+// stdout or stderr. Use the io.Writer instead.
+func exampleSnippet(w io.Writer, projectId, filePath string) {
```
[the remark this answers is not in the document]
Comment (kurtisvg) [reply content/en/_index.md:188]: So for Java, we set and capture the equivalent of `fmt.Println` and just reset it once the test is over. Could we standardize Go doing that as well? 

The reason the og working group decided to try to limit/ban return values/errors was because it was a common pattern to just return something with no context. While it's easier to test, it's harder to run and understand how to interact with the response for the user.

[the remark this answers is not in the document]
Comment (enocom) [reply content/en/_index.md:188]: I agree the writer pattern isn't great and would prefer to just return values. I also understand that can produce unintuitive samples, too.

We're moving into hacky territory by trying to capture `fmt.Println`, but we could do something like this:

``` golang
package main

import (
	"fmt"
	"log"
	"os"
)

func init() {
	f, err := os.OpenFile("test.log", os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0755)
	if err != nil {
		log.Fatal(err)
	}
	os.Stdout = f
}

func main() {
	fmt.Println("Hello there")
}
```

[the remark this answers is not in the document]
Comment (enocom) [reply content/en/_index.md:188]: See https://stackoverflow.com/a/10476304/927514 for example.

[the remark this answers is not in the document]
Comment (enocom) [reply content/en/_index.md:188]: In this approach, we'd be able to remove the io.Writer param, which would be quite nice.

[the remark this answers is not in the document]
Comment (kurtisvg) [reply content/en/_index.md:188]: IMO, using a PIPE to a buffer is clearer, and this can be implemented on the test side (without showing it in the sample).

[the remark this answers is not in the document]
Comment (codyoss) [reply content/en/_index.md:188]: I do think that passing a writer is better than overriding per test. Doing this is racy and means you can't run tests in parallel

[the remark this answers is not in the document]
Comment (enocom) [reply content/en/_index.md:188]: That seems to be the simplest approach, albeit strange by Go standards.

[the remark this answers is not in the document]
Comment (kurtisvg) [reply content/en/_index.md:188]: Do we run our tests in parallel now?

File: content/en/_index.md (hunk 5 above)
Comment (grayside) [reply content/en/_index.md:142]: My understanding the runner function itself is not part of the sample, and this is really about how we document parameters in a way that keeps us from needing to set up samples in global scope.

File: content/en/_index.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (codyoss) [reply content/en/_index.md:188]: Yes, we do have samples that test in parallel today.

[the remark this answers is not in the document]
Comment (grayside) [reply content/en/_index.md:188]: Should we use logger instead? Then we can use an instance per sample.

[the remark this answers is not in the document]
Comment (enocom) [reply content/en/_index.md:188]: io.Writer might be nicer simply because it doesn't support things like Fatalf, etc.

File: content/en/_index.md (hunk 4 above)
Comment (enocom) [reply content/en/_index.md]: Resolving this to consolidate the discussion on this thread: https://github.com/GoogleCloudPlatform/samples-style-guide/pull/43#pullrequestreview-917877148.

File: content/en/_index.md [hunk 8]
```diff
@@ -146,6 +165,14 @@ public static void main(String[] args) {
     exampleSnippet(projectId, filePath);
 }
 {{< /tab >}}
+{{< tab header="Go" >}}
+func main() {
```
Comment (enocom) [inline content/en/_index.md:169]: Go samples have typically lived in non-main packages. For this to work, we'll have to add a new package for every set of samples. Is that the expectation?

File: content/en/_index.md [hunk 9]
```diff
@@ -478,6 +583,14 @@ try {
   // IllegalArgumentException's are thrown when an invalid argument has been passed to a function. Ok to ignore.
 }
 {{< /tab >}}
+{{< tab header="Go" >}}
+// If the sample can run into errors, write the error message to the provided
+// io.Writer. Don't call log.Fatal or friends.
+if err != nil {
+    fmt.Fprintf(w, "foo.NewClient: %v", err)
```
Comment (enocom) [inline content/en/_index.md:590]: If we follow the text above, this is what the pattern becomes in Go, which is not idiomatic Go, unlike throwing exceptions in Java, Python, Ruby, and NodeJS.

Comment (enocom) [reply content/en/_index.md:590]: If the idea is that we alert users to doing something with the error, maybe this is OK, but typically, Go developers would return an error.

**Comment by enocom** [comment]
Updated this PR to address comments.

The biggest problem here is that the guide seems to be written for languages with exceptions, and doesn't account for Go wanting to return errors. Perhaps just writing an error message to stdout is adequate?

File: content/en/_index.md (hunk 8 above)
Comment (kurtisvg) [reply content/en/_index.md:169]: This was made java specific in #41 until we can reach a broader consensus on how to handle this in each language. You can remove it from this pr.

File: content/en/_index.md (hunk 9 above)
Comment (kurtisvg) [reply content/en/_index.md:590]: My 2cents is that logging the error to console is what we want to do here. Logging the error as opposed to just returning it makes it easier to run on the user.

File: content/en/_index.md (hunk 5 above)
Comment (enocom) [reply content/en/_index.md]: This was made a Java-only requirement. See https://github.com/GoogleCloudPlatform/samples-style-guide/pull/43#discussion_r839008433.

File: content/en/_index.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (enocom) [reply content/en/_index.md:210]: Consolidating this again on https://github.com/GoogleCloudPlatform/samples-style-guide/pull/43#discussion_r838999192.

File: content/en/_index.md (hunk 9 above)
Comment (enocom) [reply content/en/_index.md:531]: I've thought about this as well and come to the same conclusion. Go developers know how to return errors. I think the idea of the samples here is more about demonstrating usage. I've added a `TODO(developer): handle error` just below to communicate that. Otherwise, writing to the error message to the `io.Writer` will communicate what went wrong, so that when people copy-paste and run the sample, they get clear feedback on what went wrong.

@codyoss You're one of the Go stakeholders here. How does this sound?

File: content/en/_index.md [hunk 10]
```diff
@@ -378,6 +462,20 @@ try (CloudClient dlp = CloudClient.create()) {
...
 {{< /tab >}}
+{{< tab header="Go" >}}
+// Initialize client that will be used to send requests. This client only needs
+// to be created once, and can be reused for multiple requests. After completing
+// all of your requests, call the "Close" method on the client to safely clean
+// up any remaining background resources.
+ctx := context.Background()
+c, err := foo.NewClient(ctx)
+if err != nil {
+    // TODO(developer): handle error
```
Comment (kurtisvg) [inline content/en/_index.md:473]: I think we should update these TODOs to match the guidance on writing out output as well

File: content/en/_index.md (hunk 9 above)
Comment (kurtisvg) [reply content/en/_index.md:531]: This looks fine to me - my two nits are:

1. I think we should lead with the comment to clarify the user should update this section as needed. 
2. I think we should push for more clear (even if a bit more verbose) messages  since these are samples. 

e.g:

```
if err != nil {
    // TODO(developer): return or handle error as necessary  
    fmt.Fprintf(w, "failed to initialize foo.NewClient: %v", err)
   return
}
```

**Review by kurtisvg — APPROVED** [review]

File: content/en/_index.md (hunk 10 above)
Comment (enocom) [reply content/en/_index.md]: Done.

File: content/en/_index.md [hunk 11]
```diff
@@ -178,14 +203,22 @@ public static void exampleSnippet(String projectId, String filePath) {
     // Snippet content ...
 }
 {{< /tab >}}
+{{< tab header="Go" >}}
+// exampleSnippet demonstrates best practices.
+// The function prints to an io.Writer for testing purposes. Do not print to
+// stdout or stderr. Use the io.Writer instead.
+func exampleSnippet(w io.Writer, projectId, filePath string) {
```
Comment (grayside) [inline content/en/_index.md:210]: Should we show how to use `io.Writer` for go learners? Seems like one line that would save beginners a couple minutes looking it up and second guessing.

**Review by grayside — APPROVED** [review]
LGTM with one optional improvement comment

File: content/en/_index.md (hunk 9 above)
Comment (codyoss) [reply content/en/_index.md]: If we want to log the errors I guess that is fine, but for Go I think we should definitely still return the error as well as a part of the signature. The rest of the programing languages we support operate on an exception paradigm(I believe), and Go does not. Errors are just values and Go and should be handled. I think if we had samples that did not return errors that sets a bad precedence for our users. If this controversial take I am happy to take it offline and we can chat in a better forum. But I have a strong option here that we should return errors.

Comment (kurtisvg) [reply content/en/_index.md]: Here's my elevator pitch of trying to convince you why this rule was made in the OG 2019 samples working group. Happy to set up a time to chat as well if a higher bandwidth forum is needed. 

I do not think this is radically different than what is happening in other languages. Java specifically has 2 categories of exceptions: checked and unchecked exceptions. There are equivalent to a return value (it has to be added to the function signature) or panic (not in the function signature). 

Based on UX testing, one of the biggest goals was to create samples that are "copy-paste-runnable". The reason the rule came about around forbidding return values is because there was a pattern of snippets that create some object and then return it to be inspected for tests. These were less useful to users because it often showed the initialization of the object but not actually how to use the object in any meaningful way. It also made it harder to tell if it ran successfully, because now the user has to do something with those values to actually tell if the sample was successful or not. So the rule was created to "don't return, output to console" to force the result of the function (successful or otherwise).

I agree with you printing errors to console is not idiomatic to Go (it's not idiomatic to any of the languages). I also agree that we should be try to be handling errors in the samples (when it makes sense). However, returning an error is not handling. It's just moving the handling of the error to a different part of the code, likely not in the snippet (outside of what the user sees). But sometimes there isn't anything in the program to handle for errors - the application needs to crash and the user needs to take some non-programmatic action outside of this (enable API, change file permissions, etc). By putting a comment and writing the error to console, we are doing the following: 
1. The TODO makes it clear this isn't an idiomatic practice, and the developer should do what makes sense for their application
2. It helps highlight the cases where the error _should_ be handled in the sample 
3. It streamlines running the sample by forcing all results to the same place - the user can expect to see _something_ in the output from the function and can't just ignore the return value (accidently or otherwise) 
4. It also forces the test to actually check the output of the sample, which is the same way users are determining if the function is successful or not. 

So in conclusion, I don't think this pattern is perfect, but I think it's better than just returning values to somewhere else because it encourages better samples overall.

Comment (codyoss) [reply content/en/_index.md]: I see in one of the Java examples in this repo that is has a throws clause. To me this is similarish to returning an error in Go as we are asking the caller to due something about it. But returning an error is much lighter weight than declaring a function can throw an error. The code is just as copy-pastable if we return an error -- the caller does not even have to create a variable to capture the error or do anything with it in Go. I understand this stance of printing to show how to use for API responses, but to me errors are different, and if we are not doing any introspection with them we might as well return them as well. It makes the code more "production ready" from the get-go.

Comment (kurtisvg) [reply content/en/_index.md]: > I see in one of the Java examples in this repo that is has a throws clause.

[Here is the specific rule for Java](https://github.com/GoogleCloudPlatform/java-docs-samples/blob/main/SAMPLE_FORMAT.md#exception-handling), which here is an exception that we only let low level environment/hardware errors in the signature (again, largely because historically Java has done a poor job with Runtime Exceptions vs checked Exceptions). It specifically says all other errors should be handled or logged in the sample. 

> The code is just as copy-pastable if we return an error -- the caller does not even have to create a variable to capture the error or do anything with it in Go

A key part is samples should be copy-paste-_runnable_. Ideally samples should be easy to run as well, and leaving a return type up to the user means that could just ignore/miss it and not have feedback on why the sample failed. Returning the error means they just have to write the `if err != nil fmt.Println(err)` themselves. (Tangentially, this is why I think it's a bad idea to use a writer over fmt.Print - it's just adding extra layers to actually running the sample - but that's another argument)

> It makes the code more "production ready" from the get-go.

This isn't production code. It's meant to be instructional. We already follow a bunch of practices that don't happen in production code -  we don't typically create a new client for each request or use a writer to output content either. IMO, the comment disclaimer here clarifies that this isn't a production practice. This is why we put a comment on client initialization as well.

Comment (enocom) [reply content/en/_index.md]: For what it's worth, my own view is that returning errors is a well-established pattern that any reasonably competent Go developer knows well. So I assume anyone seeing an error printed to stdout with a note above it that says "handle or return error," will understand that the sample is about demonstrating usage and will adjust the error handling to their own needs.

cc @tritone

File: content/en/_index.md (hunk 11 above)
Comment (enocom) [reply content/en/_index.md:210]: That's a good idea. I'll update the doc string.

File: content/en/_index.md (hunk 9 above)
Comment (tritone) [reply content/en/_index.md]: I'll +1 @codyoss 's response here. Not having an error returned in Go is, well, strange. It's easier to start adapting a function from a sample code to a production use if the signature is closer to what will be correct, and it also makes the sample code more testable. Also, languages other than Java (e.g. python) will always bubble up errors as exceptions unless there is a try/except block in every sample, which I sincerely hope we're not planning on.

I also question whether the style guide needs to take a firm stance on this one way or the other. What is the cost of having this be up to whoever is implementing the samples, especially given that existing samples generally return an error and we don't plan on migrating them to follow the guide?

File: content/en/_index.md [hunk 12]
```diff
@@ -378,6 +462,22 @@ try (CloudClient dlp = CloudClient.create()) {
   // make a request with the client
 }
 {{< /tab >}}
+{{< tab header="Go" >}}
+// Initialize client that will be used to send requests. This client only needs
```
Comment (tritone) [inline content/en/_index.md:466]: IMO this is way too wordy to include in every sample.

File: content/en/_index.md [hunk 13]
```diff
@@ -178,14 +203,22 @@ public static void exampleSnippet(String projectId, String filePath) {
     // Snippet content ...
 }
 {{< /tab >}}
+{{< tab header="Go" >}}
+// exampleSnippet demonstrates best practices.
+// The function prints to an io.Writer for testing purposes. Do not print to
```
Comment (tritone) [inline content/en/_index.md:208]: Just confirming that this comment is something to help the sample implementer and not something to be included in the samples themselves.

File: content/en/_index.md [hunk 14]
```diff
@@ -275,22 +308,66 @@ async function listInfoTypes() {
...
+		// Defaults to "supported_by=INSPECT"
+		Filter: "supported_by=INSPECT",
+		// BCP-47 language code for localized infoType friendly names.
+		// Defaults to "en_US"
+		LanguageCode: "en-US",
+	}
+
+	resp, err := c.ListInfoTypes(ctx, req)
+	if err != nil {
+		// TODO: Handle error.
```
Comment (tritone) [inline content/en/_index.md:346]: This needs a return statement even if we are not returning errors. Same in L329 above

File: content/en/_index.md [hunk 15]
```diff
@@ -275,22 +308,66 @@ async function listInfoTypes() {
...
+	// needs to be created once, and can be reused for multiple requests. After
+	// completing all of your requests, call the "Close" method on the client to
+	// safely clean up any remaining background resources.
+	c, err := dlp.NewClient(ctx)
+	if err != nil {
+		// TODO: Handle error.
+	}
+	defer c.Close()
+
+	// Construct the request to be sent by the client
```
Comment (tritone) [inline content/en/_index.md:333]: This line maybe feels helpful for GAPIC/Apiary but is overkill for a manual client IMO

File: content/en/_index.md [hunk 16]
```diff
@@ -275,22 +308,66 @@ async function listInfoTypes() {
...
+		// Only return infoTypes supported by certain parts of the API.
+		// Supported filters are "supported_by=INSPECT" and "supported_by=RISK_ANALYSIS"
+		// Defaults to "supported_by=INSPECT"
+		Filter: "supported_by=INSPECT",
+		// BCP-47 language code for localized infoType friendly names.
+		// Defaults to "en_US"
+		LanguageCode: "en-US",
+	}
+
+	resp, err := c.ListInfoTypes(ctx, req)
```
Comment (tritone) [inline content/en/_index.md:344]: Another Go-specific thing is that in most libraries, requests will not time out unless the context has a deadline. For storage, we got enough customer complaints about hanging calls etc that I ended up adding setting a deadline to every sample.

File: content/en/_index.md (hunk 12 above)
Comment (kurtisvg) [reply content/en/_index.md:466]: So if you have suggestions to make this less terse we can talk about it, but this is a significant issue with our current samples. The GKE team escalated concerns about a bunch of users taking the recommendations too literally: [go/reuse-cloud-client](http://go/reuse-cloud-client)

File: content/en/_index.md (hunk 16 above)
Comment (kurtisvg) [reply content/en/_index.md:344]: I think this is a great suggestion, maybe we can follow it up in a different PR under the language-specific stuff.

File: content/en/_index.md (hunk 15 above)
Comment (kurtisvg) [reply content/en/_index.md:333]: This is in the current canonical: https://github.com/GoogleCloudPlatform/golang-samples/blob/68e04de0b13bfad4d34c9556ca66ed40655dcdd4/dlp/snippets/inspect/inspect_string.go#L40

File: content/en/_index.md (hunk 13 above)
Comment (enocom) [reply content/en/_index.md:234]: Yes, that's correct.

**Comment by enocom** [comment]
There are a few open issues with this PR:

> 1. How these standards will affect the large number of existing samples. Do we migrate to this standard? Who does that work?

This is worth discussing, but is out of scope for this PR. I think we need to agree to a standard here for new samples and then decide how to handle what's already out there. I assume other languages have the same problem. 

With the exception of error handling, the changes here are not a significant departure from existing Go samples.

> 2. Are the samples meant to be production-quality?

I don't see any mention in the existing [CONTRIBUTING](https://github.com/GoogleCloudPlatform/golang-samples/blob/main/CONTRIBUTING.md), but have heard people say that the existing samples are meant to be production quality. This sample guide here takes a different approach and instead focuses on making the samples instructional.

> 3. Do we return an error in sample functions, write to the io.Writer, or both?

Here are the two competing options:

``` golang
// with an error return
func exampleSnippet(w io.Writer) error {
  // ...
  if err != nil {
    return err  
  }
  // ...
}
```

Versus:

``` golang
// with writing the error to the io.Writer
func exampleSnippet(w io.Writer) {
  // ...
  if err != nil {
    // TODO(developer): return or handle error as necessary
    fmt.Fprintf(w, "uh-oh: %v", err)
    return  
  }
  // ...
}
```

In my understanding, the workflow we're designing for is copy, paste into a main func, and then run it to see what happens. In this world, the error return isn't as important as seeing how the pieces fit together. I assume adapting a snippet to production code would involve:

1. Moving the client initialization to a more central place where it could be shared
2. Handling errors as needed, changing function signatures (getting rid of the io.Writer is obviously a necessity)
3. Breaking the snippet into separate calls (assuming the developer wants to do a few things with the client).

In other words, returning or not returning the error is tangential in how the sample would be used.

File: content/en/_index.md (hunk 12 above)
Comment (grayside) [reply content/en/_index.md:539]: It never occurred to me all the comments were meant to be copied verbatim. That might be something we need to make more clear. I think there are some opportunities to make this more concise.

1. Reduce the client instantiation comment to:
```
// Initialize a client that will be used to send requests.
// This client can be created once and reused for multiple requests.
```

Move the following above the defer call:

```
// Call the 'Close' method on the client to safely clean up background resources.
```

File: content/en/_index.md (hunk 16 above)
Comment (grayside) [reply content/en/_index.md:383]: +1 to adding new rules later, an considering if that rule should be extended into the examples for other rules.

File: content/en/_index.md (hunk 9 above)
Comment (grayside) [reply content/en/_index.md]: Removing established rules is not in scope right now, but it's clear this is one of the rules that hits one of Go's strongest language opinions. I think we should go to the goals of the rule, figure out what's needed, then consider if the rule's language might need revisions.

The main goal as I understand is to make sure that **developers handle errors and avoid failures that a learner can't distinguish from a product error**

With that goal in mind, and in the background other concerns about copy-paste-runnable and cross-language consistency, I think it's fair to reframe the strict expectations we have in other languages.

Looking at this again, my thoughts are:
* Returning an error in go is idiomatic, not returning an error is almost never ok
* For Go developers, returning the error does not seem like an excessive distraction
* Not sure why I'd log an error to an io.Writer instead of, well, logging the error. This makes sense to me for catching function output, not errors
* We shouldn't assume sample readers or authors are experts in Go, programming, or Cloud.

Could we remind folks to log, have the copied sample log by default, and idiomatically return the error with something like this?

```
// If the sample can run into errors, write the error message to the provided
if err != nil {
    // TODO(developer): log this error from the calling function.
    log.Println("failed to initialize foo.Client: %v", err)
    return err
}
```

If this proposal isn't getting us closer to a resolution, let's plan to talk this through.

Comment (enocom) [reply content/en/_index.md]: To summarize an offline conversation with @tritone, @codyoss, @kurtisvg, we're going to keep the error return for now. At some future date, we will remove the io.Writer, work that depends on a clear testing strategy.

File: content/en/_index.md (hunk 12 above)
Comment (enocom) [reply content/en/_index.md:599]: Done.

**Comment by enocom** [comment]
@codyoss @tritone This is ready for another pass.

**Comment by enocom** [comment]
My PR doesn't have permission to push to the preview site.

File: content/en/_index.md [hunk 17]
```diff
@@ -362,6 +397,52 @@ async function listInfoTypes() {
...
+	"fmt"
+
+	dlp "cloud.google.com/go/dlp/apiv2"
+	dlppb "google.golang.org/genproto/googleapis/privacy/dlp/v2"
+)
+
+// listInfoTypes lists the types of sensitive information the DLP API supports.
+func listInfoTypes(w io.Writer) error {
+	ctx := context.Background()
+    // Initialize a client that will be used to send requests.
```
Comment (codyoss) [inline content/en/_index.md:412]: I know this was talked about earlier today, did people have a preference if we group all the client comments together in a more concise way. I think this is fine for now and could be iterated in the future though too. One thing that we might want to rephrase this so people know that they can use clients across goroutines. I think that implies reuse and answers a common question I have seen in the past.

**Review by codyoss — APPROVED** [review]
Just one minor thing, or else it LGTM

File: content/en/_index.md (hunk 17 above)
Comment (enocom) [reply content/en/_index.md]: I agree about tightening up the wording and mentioning thread safety. I've condensed the comments to:

```
// Initialize a client once and reuse it to send multiple requests. Clients
// are safe to use across goroutines. When the client is no longer needed,
// call the Close method to cleanup its resources.
```

Comment (kurtisvg) [reply content/en/_index.md]: I'm +1 to mentioning thread safety. Current version LGTM

**Review by tritone — APPROVED** [review]
Thanks for your work on this Eno!
